hw/hw14/election.py

```diff
@@ -0,0 +1,333 @@
+    output = []
+    for row in csv.DictReader(open(path)):
+        output.append(row)
+    return output
+
+
+################################################################################
+# Problem 1: State edges
+################################################################################
+
+def row_to_edge(row):
+    """
+    Given an *ElectionDataRow* or *PollDataRow*, returns the
+    Democratic *Edge* in that *State*.
+    """
+    return float(row["Dem"]) - float(row["Rep"])
+
+def state_edges(election_result_rows):
+    """
+    Given a list of *ElectionDataRow*s, returns *StateEdge*s.
+    The input list has no duplicate *States*;
+    that is, each *State* is represented at most once in the input list.
+    """
+    d = {}
+    for row in election_result_rows:
+        election_result_row = row
+        state = election_result_row["State"]
+        edge = row_to_edge(row)
+        d[state] = edge
+    return d
+
+
+################################################################################
+# Problem 2: Find the most recent poll row
+################################################################################
+
+def earlier_date(date1, date2):
+    """
+    Given two dates as strings (formatted like "Oct 06 2012"), returns True if
+    date1 is before date2.
+    """
+    return (time.strptime(date1, "%b %d %Y") < time.strptime(date2, "%b %d %Y"))
+
+def most_recent_poll_row(poll_rows, pollster, state):
+    """
+    Given a list of *PollDataRow*s, returns the most recent row with the
+    specified *Pollster* and *State*. If no such row exists, returns None.
+    """
+    most_recent = {}
+    for row in poll_rows:
+        if row["Pollster"] == pollster and row["State"] == state:
+            if most_recent == {}:
+                most_recent = row
+            elif earlier_date(most_recent["Date"], row["Date"]):
+                most_recent = row
+    if most_recent == {}:
+        return None
+    else:
+        return most_recent
+
+
+################################################################################
+# Problem 3: Pollster predictions
+################################################################################
+
+def unique_column_values(rows, column_name):
+    """
+    Given a list of rows and the name of a column (a string),
+    returns a set containing all values in that column.
+    """
+    s = set()
+    for row in rows:
+        s.add(row[column_name])
+    return s
+
+
+def pollster_predictions(poll_rows):
+    """
+    Given a list of *PollDataRow*s, returns *PollsterPredictions*.
+    For a given pollster, uses only the most recent poll for a state.
+    """
+    d = {}
+    pollsters = unique_column_values(poll_rows, "Pollster")
+    states = unique_column_values(poll_rows, "State")
+    for pollster in pollsters:
+        d[pollster] = {}
+        for state in states:
+            recent = most_recent_poll_row(poll_rows, pollster, state)
+            if recent is not None:
```

Comment:
You can also write this as `if recent:`. This will evaluate to true as long as recent is not None.